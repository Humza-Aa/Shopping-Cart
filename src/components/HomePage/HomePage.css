html,
body,
.App,
#root {
  margin: 0;
  height: 100%;
}

.test {
  height: 140vh;
}

.homePage {
  padding: 10px;
  display: flex;
  flex-wrap: wrap;
  gap: 10px 0;
  justify-content: space-evenly;
  background-color: rgb(225, 225, 225);
  height: 100%;
}

.itemCard {
  background-color: rgb(255, 255, 255);
  height: 290px;
  width: 220px;
  border-radius: 5px;
  border: 1px solid rgb(225, 225, 225);
}

.infoDiv {
  border-top: 2px solid rgb(225, 225, 225);
  display: flex;
  flex-direction: column;
  padding: 10px;
  gap: 10px;
}

.nameDiv {
    min-height: 36px;
}

.imgDiv {
  height: 150px;
  width: 100%;
  display: flex;
  justify-content: center;
}

.imgDiv > img {
  max-width: 100%;
  max-height: 100%;
  border-radius: 10px;
}

.nameDiv {
  display: flex;
  align-items: center;
}

p {
  margin: 0;
}

.addCartBtnDiv > button {
  width: 100%;
  height: 30px;
  background-color: aquamarine;
}
