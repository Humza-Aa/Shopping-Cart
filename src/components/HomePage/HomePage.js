import React from "react";
import "./HomePage.css";

const HomePage = (itemlist) => {
  
  return (
    <div className="homePage">
      {itemlist.itemNames.map((item) => {
        return (
          <div key={item.key} className="itemCard">
            <div className="imgDiv">
              <img
                src= {item.imgSrc}
                alt="Item"
                data-testid="itemImage"
              />
            </div>
            <div className="infoDiv">
              <div className="nameDiv">
                <p data-testid='itemName'>{item.name}</p>
              </div>
              <div className="priceDiv">
                <p>${item.price}</p>
              </div>
              <div className="addCartBtnDiv">
                <button data-testid="button" onClick={(e) => {itemlist.addItem(item.key)}}>Add to Cart</button>
              </div>
            </div>
          </div>
        );
      })}
    </div>
  );
};

export default HomePage;
