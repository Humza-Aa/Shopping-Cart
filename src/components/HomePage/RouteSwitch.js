import React, { useEffect } from "react";
import { BrowserRouter, Routes, Route } from "react-router-dom";
import HomePage from "./HomePage";
import NavBar from "../NavBar/NavBar";
import ShoppingCart from "../ShoppingCart/ShoppingCart";
import { useState } from "react";
import { doc, updateDoc, onSnapshot, collection } from "firebase/firestore";
import { auth, db } from "../../firebase-config";
import { onAuthStateChanged } from "firebase/auth";

const RouteSwitch = () => {
  const [shoppingList, setshoppingList] = useState([]);
  const [itemQuantity, setitemQuantity] = useState({
    0: 0,
    1: 0,
    2: 0,
    3: 0,
    4: 0,
    5: 0,
  });

  const [itemNames] = useState([
    {
      name: "Dell Inspiron 3000 15.6 Touchscreen Laptop - Black",
      imgSrc:
        "https://multimedia.bbycastatic.ca/multimedia/products/500x500/159/15972/15972303.jpg",
      price: 600,
      key: 0,
      quantity: 0,
    },
    {
      name: "HP 15.6 Laptop - Jet Black",
      imgSrc: "https://m.media-amazon.com/images/I/71yZp-Ru4WL._AC_SX679_.jpg",
      price: 550,
      key: 1,
      quantity: 0,
    },
    {
      name: "Microsoft Surface Laptop Go 12.4 Multi-Touch, Sand Stone",
      imgSrc: "https://m.media-amazon.com/images/I/61VcLC0G13L._AC_SL1500_.jpg",
      price: 430,
      key: 2,
      quantity: 0,
    },
    {
      name: "Acer Nitro 5 15.6 Gaming Laptop - Black",
      imgSrc:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTv-36ZG-XQuCNUEkc5XItH25qxrkSkPY4Kfg&usqp=CAU",
      price: 850,
      key: 3,
      quantity: 0,
    },
    {
      name: "ASUS ROG Strix G15 15.6 Gaming Laptop",
      imgSrc: "https://m.media-amazon.com/images/I/71ZTl79+18L._AC_SX569_.jpg",
      price: 1200,
      key: 4,
      quantity: 0,
    },
    {
      name: "Acer ROG Strix G15 15.6 Gaming Laptop",
      imgSrc: "https://m.media-amazon.com/images/I/71ZTl79+18L._AC_SX569_.jpg",
      price: 1200,
      key: 5,
      quantity: 0,
    },
  ]);

  const [currentUser, setCurrentUser] = useState({});
  const [dbItemQuantity, setDbItemQuantity] = useState([]);

  useEffect(() => {
    onAuthStateChanged(auth, (user) => {
      setCurrentUser(user);
    });
  }, []);

  useEffect(
    () =>
      onSnapshot(collection(db, "itemQuantities"), (snapshot) =>
        setDbItemQuantity(
          snapshot.docs.map((doc) => ({ itemQ: {...doc.data()}, id: doc.id }))
        )
      ),
    []
  );

  const addItem = (itemIndex) => {
    setshoppingList([...shoppingList, itemIndex]);
    setitemQuantity({
      ...itemQuantity,
      [itemIndex]: itemQuantity[itemIndex] + 1,
    });
  };

  const saveData = async (itemIndex) => {
    console.log(currentUser.displayName);
    const docRef = doc(db, "itemQuantities", currentUser.userName);
    // const payLoad = itemQ;
    // await setDoc(docRef, payLoad);
    await updateDoc(docRef, {
      [itemIndex]: itemQuantity[itemIndex] + 1,
    });
  };

  return (
    <BrowserRouter>
      <NavBar
        shoppingList={shoppingList}
        setshoppingList={setshoppingList}
        setCurrentUser={setCurrentUser}
        currentUser={currentUser}
      />
      <Routes>
        <Route
          path="/"
          element={<HomePage itemNames={itemNames} addItem={addItem} />}
        />
        <Route
          path="/shoppingcart"
          element={
            <ShoppingCart
              itemNames={itemNames}
              shoppingList={shoppingList}
              itemQuantity={itemQuantity}
              dbItemQuantity={dbItemQuantity}
              currentUser={currentUser}
            />
          }
        />
      </Routes>
    </BrowserRouter>
  );
};

export default RouteSwitch;
