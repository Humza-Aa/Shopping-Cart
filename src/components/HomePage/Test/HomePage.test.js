import React from "react";
import "@testing-library/jest-dom";
import HomePage from "../HomePage";
import renderer from 'react-test-renderer'
import userEvent from "@testing-library/user-event";
import { render, cleanup, screen} from "@testing-library/react";

describe("Testing Item Display Div", () => {
    let itemNames = [];
    beforeEach(() => {
        itemNames = [{
            name: "Microsoft Surface Laptop Go 12.4 Multi-Touch, Sand Stone",
            imgSrc: "https://m.media-amazon.com/images/I/61VcLC0G13L._AC_SL1500_.jpg",
            price: 430,
            key: 2,
            quantity: 0,
        }];
    });

    afterEach(cleanup);

    it('renders correctly', () => {
        const tree = renderer
          .create(<HomePage itemNames={itemNames}></HomePage>)
          .toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    it("Image correctly loaded", () => {
        const {getByTestId} = render(<HomePage itemNames={itemNames}></HomePage>)
        const imageSrc = getByTestId('itemImage').getAttribute('src')
        expect(imageSrc).toEqual('https://m.media-amazon.com/images/I/61VcLC0G13L._AC_SL1500_.jpg');
    });
    it("Name correctly loaded", () => {
        const {getByTestId} = render(<HomePage itemNames={itemNames}></HomePage>)
        const itemName = getByTestId('itemName');
        expect(itemName).toHaveTextContent('Microsoft Surface Laptop Go 12.4 Multi-Touch, Sand Stone')
    });
})

describe("Test Shopping Cart Quantity", () => {
    let itemNames = [];
    beforeEach(() => {
        itemNames = [{
            name: "Microsoft Surface Laptop Go 12.4 Multi-Touch, Sand Stone",
            imgSrc: "https://m.media-amazon.com/images/I/61VcLC0G13L._AC_SL1500_.jpg",
            price: 430,
            key: 2,
            quantity: 0,
        }];
    });
    it("Increasing Quantity", () => {
        const shoppingCartLenght = jest.fn();
        render(<HomePage itemNames={itemNames} addItem={shoppingCartLenght}></HomePage>)
        const button = screen.getByTestId('button');
        userEvent.dblClick(button);
        userEvent.dblClick(button);
        expect(shoppingCartLenght).toHaveBeenCalledTimes(4)
    });
});
