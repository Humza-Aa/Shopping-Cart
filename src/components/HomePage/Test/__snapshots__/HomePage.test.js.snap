// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Item Display Div renders correctly 1`] = `
<div
  className="homePage"
>
  <div
    className="itemCard"
  >
    <div
      className="imgDiv"
    >
      <img
        alt="Item"
        data-testid="itemImage"
        src="https://m.media-amazon.com/images/I/61VcLC0G13L._AC_SL1500_.jpg"
      />
    </div>
    <div
      className="infoDiv"
    >
      <div
        className="nameDiv"
      >
        <p
          data-testid="itemName"
        >
          Microsoft Surface Laptop Go 12.4 Multi-Touch, Sand Stone
        </p>
      </div>
      <div
        className="priceDiv"
      >
        <p>
          $
          430
        </p>
      </div>
      <div
        className="addCartBtnDiv"
      >
        <button
          data-testid="button"
          onClick={[Function]}
        >
          Add to Cart
        </button>
      </div>
    </div>
  </div>
</div>
`;
